neqo-transport/src/connection/tests/datagram.rs

```diff
@@ -224,6 +225,76 @@ fn datagram_acked() {
     ));
 }
 
+fn send_packet_and_check_server_event(
+    client: &mut Connection,
+    server: &mut Connection,
+    event: &ConnectionEvent,
+) {
+    let out = client.process_output(now()).dgram();
+    server.process_input(out.unwrap(), now());
+    let events: Vec<_> = server
+        .events()
+        .filter_map(|evt| match evt {
+            ConnectionEvent::RecvStreamReadable { .. } | ConnectionEvent::Datagram { .. } => {
+                Some(evt)
+            }
+            _ => None,
+        })
+        .collect();
+    // We should only get one event - either RecvStreamReadable or Datagram.
+    assert_eq!(events.len(), 1);
+    assert_eq!(mem::discriminant(event), mem::discriminant(&events[0]));
+}
+
+#[test]
+fn datagram_after_stream_data() {
+    let (mut client, mut server) = connect_datagram();
+
+    // Write a datagram first.
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_datagram(DATA_MTU, Some(1)), Ok(()));
+
+    // Create a stream with normal priority and send some data.
+    let stream_id = client.stream_create(StreamType::BiDi).unwrap();
+    client.stream_send(stream_id, &[6; 1200]).unwrap();
+
+    let desired_event = ConnectionEvent::RecvStreamReadable { stream_id };
+    send_packet_and_check_server_event(&mut client, &mut server, &desired_event);
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent);
+
+    let desired_event = ConnectionEvent::Datagram([].to_vec());
```

Comment:
If you do as I suggest, you can do `matches!(..., ConnectionEvent::Datagram(_))` instead.  Or even...

```rust
   if let ConnectionEvent::Datagram(data) = send_packet_and_get_server_event(&mut client, &mut server) {
      assert_eq!(data, DATA_MTU);
   } else {
      panic!();
   }
```